contributors/design-proposals/daemonset-create-first-rolling-update.md

```diff
@@ -0,0 +1,178 @@
+## DaemonSet create-first rolling update strategy
+
+DaemonSets currently support a "delete first, then add" rolling update
+strategy.
+
+This document presents a design for a new "add first, then delete" rolling
+update strategy, called `SurgingRollingUpdate`. This strategy will create new
+pods first, cleaning up the old pods once the new ones are running. As with
+Deployments, users will use a `maxSurge` parameter to control how many nodes
+are allowed to have >1 pod running during a rolling update.
+
+### Example use case
+
+For self-hosted Kubernetes DaemonSets are a natural fit for running the
+scheduler and controller-manager on master nodes. However, since DaemonSets are
```

Comment:
will the proposed approach even work for scheduler and controller manager the way kubeadm wants it to? running with hostNetwork, they will try to bind their ports for health checks and be blocked by the existing pods, right?

Reply:
Sounds like any host resources (like hostPath) will have conflicts.

Reply:
so, it doesn't sound like implementing this will unblock kubeadm's use of daemonsets for single master self-hosting

Reply:
We could bind the second time with SO_REUSEADDR to get around the hostPort issue.

HostPath will conflict but wouldn't likely complain and could easily result in some form of data corruption.

Reply:
In `bootkube` we don't run the scheduler or controller-manager with host networking. The apiserver does, but we already run that as a daemonset and use the standard kill-then-start strategy (which the checkpointer helps facilitate) during upgrades.

I'm not sure what the plan is for `kubeadm` however. cc @luxas